src/crystal/system/unix/process.cr

```diff
@@ -127,40 +128,87 @@ struct Crystal::System::Process
     )
   end
 
+  # The RWLock around `fork` is trying to protect against file descriptors
+  # leaking to sub-processes.
+  #
+  # There is a race condition in the POSIX standard between the creation of a
+  # file descriptor (e.g. `open`, `socket`, `accept`) and setting the
+  # `FD_CLOEXEC` flag with `fcntl`. During the time window between those two
+  # syscalls, another thread may fork the process and exec another process,
+  # which will leak the file descriptor.
+  #
+  # Most systems have long implemented non standard syscalls that prevent the
+  # race condition, except for Darwin that implements `O_CLOEXEC` but doesn't
+  # implement `SOCK_CLOEXEC`, and supports neither of `accept4`, `dup3` or
+  # `pipe2`.
+  #
+  # NOTE: there may still be some potential leaks (e.g. calling `accept` on a
+  #       blocking socket).
+  {% if LibC.has_constant?(:SOCK_CLOEXEC) && LibC.has_method?(:accept4) && LibC.has_method?(:dup3) && LibC.has_method?(:pipe2) %}
+    # we don't implement #disable_fork so compilation will fail if we need to
+    # support another case, instead of silently skipping the rwlock!
+
+    def self.fork_lock(&)
+      yield
+    end
+  {% else %}
+    @@fork_rwlock = Crystal::RWLock.new
+
+    def self.disable_fork(&)
+      @@fork_rwlock.read_lock
+      begin
+        yield
+      ensure
+        @@fork_rwlock.read_unlock
+      end
+    end
+
+    def self.fork_lock(&)
+      @@fork_rwlock.write_lock
+      begin
+        yield
+      ensure
+        @@fork_rwlock.write_unlock
+      end
+    end
+  {% end %}
+
   def self.fork(*, will_exec = false)
-    newmask = uninitialized LibC::SigsetT
-    oldmask = uninitialized LibC::SigsetT
-
-    LibC.sigfillset(pointerof(newmask))
-    ret = LibC.pthread_sigmask(LibC::SIG_SETMASK, pointerof(newmask), pointerof(oldmask))
-    raise RuntimeError.from_errno("Failed to disable signals") unless ret == 0
-
-    case pid = LibC.fork
-    when 0
-      # child:
-      pid = nil
-      if will_exec
-        # reset signal handlers, then sigmask (inherited on exec):
-        Crystal::System::Signal.after_fork_before_exec
-        LibC.sigemptyset(pointerof(newmask))
-        LibC.pthread_sigmask(LibC::SIG_SETMASK, pointerof(newmask), nil)
+    fork_lock do
```

Comment:
I so wish we could at some point get to a state where `fork` could just be deleted. So much extra complexity to support a model that just aint that great.

Are these also needed for things that execute subprograms, like 
```
puts `ls`
```
?

Reply:
This is the exact use case, yes; we don't support `fork` publicly when MT is enabled, but we still need to handle fork/exec and we need to set the CLOEXEC flag.

Reply:
BTW: the issue is with `exec` but I must prevent `fork` since there won't be any way to fix the issue after the process forks.

And... I'm wondering if the same lock shouldn't prevent `exec` too :thinking:

Reply:
I checked [`posix_spawn`](https://www.man7.org/linux/man-pages/man3/posix_spawn.3.html) and it would have the same issue as it appears to be implemented with fork/exec, on Linux at least.